packages/backend-common/src/reading/integration.test.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ConfigReader } from '@backstage/config';
+import { getVoidLogger } from '../logging';
+import { UrlReaders } from './UrlReaders';
+
+const reader = UrlReaders.default({
+  logger: getVoidLogger(),
+  config: new ConfigReader({
+    // The tokens in this config provide read only access to the backstage-verification repos
+    integrations: {
+      github: [
+        {
+          host: 'github.com',
+        },
+      ],
+      gitlab: [
+        {
+          host: 'gitlab.com',
+          token: 'tveGtSHDBJM9ZRHZNRfm',
+        },
+      ],
+      bitbucket: [
+        {
+          host: 'bitbucket.org',
+          username: 'backstage-verification',
+          appPassword: 'H79MAAhtbZwCafkVTrrQ',
+        },
+      ],
+      azure: [
+        {
+          host: 'dev.azure.com',
+          // lasts until 2022-01-27
+          token: 'bhs5cbukiuxrkc3ftuyt5h3eqewtkj37lmf3jx5aoajivq3f5jmq',
+        },
+      ],
+    },
+  }),
+});
+
+function withRetries(count: number, fn: () => Promise<void>) {
+  return async () => {
+    let error;
+    for (let i = 0; i < count; i++) {
+      try {
+        await fn();
+        return;
+      } catch (err) {
+        error = err;
+      }
+    }
+    throw error;
+  };
+}
+
+describe('UrlReaders', () => {
+  it(
+    'should read data from azure',
+    withRetries(3, async () => {
+      const data = await reader.read(
+        'https://dev.azure.com/backstage-verification/test-templates/_git/test-templates?path=%2Ftemplate.yaml',
+      );
+      expect(data.toString()).toContain('test-template-azure');
+
+      const res = await reader.readTree(
+        'https://dev.azure.com/backstage-verification/test-templates/_git/test-templates?path=%2F{{cookiecutter.name}}',
```

Comment:
Azure readTree is not very smart like other providers - it doesn't work with non-default branch and it can not read from sub path. I might be wrong.

https://github.com/backstage/backstage/blob/1ac2ba6e8e1e3a44aaf85e9f20bb19a1404111da/packages/backend-common/src/reading/AzureUrlReader.ts#L79

Reply:
So a good URL for this test is `'https://dev.azure.com/backstage-verification/test-templates/_git/test-templates`.

Reply:
I want to specifically test the subpath request